hibernate-cache-local/src/main/java/org/infinispan/tutorial/simple/hibernate/cache/local/InfinispanHibernateCacheLocal.java

```diff
@@ -0,0 +1,268 @@
+      printfAssert("Query cache hit: %d (expected %d)%n", stats.getQueryCacheHitCount(), 1);
+
+      // Update one of the persisted entities, stats should show a cache hit and a cache put
+      updateEntity(2L);
+      eventCacheStats = getCacheStatistics(Event.class.getName());
+      printfAssert("Event entity cache hits: %d (expected %d)%n", eventCacheStats.getHitCount(), 1);
+      printfAssert("Event entity cache puts: %d (expected %d)%n", eventCacheStats.getPutCount(), 1);
+
+      // Repeat query after update, expect:
+      // * no cache hits or puts since entities are already cached
+      // * a query cache miss and query cache put, because when an entity is updated,
+      //   any queries for that type are invalidated
+      queryEntities();
+      stats = getStatistics();
+      printfAssert("Query cache miss: %d (expected %d)%n", stats.getQueryCacheMissCount(),1);
+      printfAssert("Query cache put: %d (expected %d)%n", stats.getQueryCachePutCount(), 1);
+
+
+      // Save cache-expiring entity, stats should show a second level cache put
+      saveExpiringEntity();
+      personCacheStats = getCacheStatistics(Person.class.getName());
+      printfAssert("Person entity cache puts: %d (expected %d)%n", personCacheStats.getPutCount(), 1);
+
+      // Find expiring entity, stats should show a second level cache hit
+      findExpiringEntity(4L);
+      personCacheStats = getCacheStatistics(Person.class.getName());
+      printfAssert("Person entity cache hits: %d (expected %d)%n", personCacheStats.getHitCount(), 1);
+
+      // Wait long enough for entity to be expired from cache
+      Thread.sleep(1100);
+
+      // Find expiring entity, after expiration entity should come from DB
+      // Stats should show a cache miss and a cache put
+      findExpiringEntity(4L);
+      personCacheStats = getCacheStatistics(Person.class.getName());
+      printfAssert("Person entity cache miss: %d (expected %d)%n", personCacheStats.getMissCount(), 1);
+      printfAssert("Person entity cache put: %d (expected %d)%n", personCacheStats.getPutCount(), 1);
+
+      // Close persistence manager
+      emf.close();
+   }
+
+   private static void persistEntities() {
+      try(Session em = createEntityManagerWithStatsCleared()) {
+         EntityTransaction tx = em.getTransaction();
+         try {
+            tx.begin();
+
+            em.persist(new Event("Caught a pokemon!"));
+            em.persist(new Event("Hatched an egg"));
+            em.persist(new Event("Became a gym leader"));
+         } catch (Throwable t) {
+            tx.setRollbackOnly();
+            throw t;
+         } finally {
+            if (tx.isActive()) tx.commit();
+            else tx.rollback();
+         }
+      }
+   }
+
+   private static Session createEntityManagerWithStatsCleared() {
+      EntityManager em = emf.createEntityManager();
+      emf.unwrap(SessionFactory.class).getStatistics().clear();
+      return em.unwrap(Session.class);
```

Comment:
I think that you should add a comment that `EntityMangager` is not `AutoCloseable` to explain why are you using a `Session` instead... `unwrap` to get stats is to-be-expected, but using it for EM looks hackish.